QMI PDC support for MCFG file downloading in the MM plugin

aleksander0m opened the pull request:
@hughsie @superm1 this is what I came up with for the DW5821e....

This module requires first a fastboot-based firmware download (e.g. to flash the modem system) plus then N qmi-pdc-based downloads (one for each carrier-specific configuration provided for the associated modem system).

The most controversial change is probably the looping in the engine and how the plugins specify the list of supported methods that the engine should trigger for separate detach/update/attach operations.

Comments welcome!

aleksander0m commented:
> This probably means more stuff in dependencies.xml too

Just pushed a fixup for contrib/ci/dependencies.xml

hughsie commented on plugins/modem-manager/fu-mm-utils.c:
```diff
@@ -0,0 +1,80 @@
+                /* Check next parent */
+                next = g_udev_device_get_parent (parent);
+                g_object_unref (parent);
+                parent = next;
```
`g_set_object (&parent, next)`?

aleksander0m replied:
doesn't that also increase the refcount of next? although then I could make next a g_autoptr, right?

aleksander0m replied:
done in a new fixup commit

hughsie commented on plugins/modem-manager/fu-qmi-pdc-updater.c:
```diff
@@ -0,0 +1,433 @@
+			return;
+		}
+
+		ctx->error = g_error_new (G_IO_ERROR, G_IO_ERROR_FAILED,
```
`g_set_error?`

aleksander0m replied:
just pushed a new fixup commit to handle this

hughsie commented on plugins/modem-manager/fu-mm-device.c:
```diff
@@ -239,68 +384,210 @@ fu_mm_device_detach_fastboot (FuDevice *device, GError **error)
+	GList		*file_infos;
+	gsize		 total_written;
+	gsize		 total_bytes;
+} ArchiveIterateContext;
```
FuMmArchiveIterateCtx perhaps?

aleksander0m replied:
pushed a new fixup to handle this renaming

hughsie commented on plugins/modem-manager/fu-mm-device.c:
```diff
@@ -239,68 +384,210 @@ fu_mm_device_detach_fastboot (FuDevice *device, GError **error)
+	file_info = g_new0 (FileInfo, 1);
+	file_info->filename = g_strdup (filename);
+	file_info->bytes = g_bytes_ref (bytes);
+	ctx->file_infos = g_list_prepend (ctx->file_infos, file_info);
```
Could we do it as a GPtrArray with a free fun defined please?

aleksander0m replied:
done in a new fixup commit in the branch

hughsie commented:
> The most controversial change is probably the looping in the engine

Yes, predictably I don't like that at all. I think we can reorganise things to make them much simpler. Can you clarify a bit what you'd like to do in an ideal world? e.g.

    modem-manager::detach()
    fastboot::write_firmware()
    fastboot::attach()
    modem-manager::write_firmware()
    modem-manager::detach()

I think we can some up with something somewhat simpler by adding API in the right places.

aleksander0m commented:
> > The most controversial change is probably the looping in the engine
> 
> Yes, predictably I don't like that at all. I think we can reorganise things to make them much simpler. Can you clarify a bit what you'd like to do in an ideal world? e.g.
> 
> ```
> modem-manager::detach()
> fastboot::write_firmware()
> fastboot::attach()
> modem-manager::write_firmware()
> modem-manager::detach()
> ```
> I think we can some up with something somewhat simpler by adding API in the right places.

The minimum flow for the DW5821e, which needs first fastboot, then qmi-pdc, would be:

- modem-manager::detach()   ---> inhibits MM modem and triggers fastboot mode
- fastboot::write_firmware() ----> downloads via fastboot
- fastboot::attach()  ---> reboots in modem mode, we detect the new FuMmDevice based on udev, not based on MM (as we're inhibited)
- modem-manager::write_firmware()  ----> downloads via qmi-pdc
- modem-manager:attach() ----> uninhibits MM modem and so it waits for MM to reprobe and re-expose it again.

Now, not all modems will need fastboot+qmi-pdc, some may need just fastboot, and some may need just qmi-pdc

hughsie commented:
> The minimum flow for the DW5821e, which needs first fastboot, then qmi-pdc

So, what about this as a general rule: _if the plugin changes during detach_....

* run NewPlugin->write_firmware()
* run NewPlugin->attach()
* run OldPlugin->write_firmware()
* run OldPlugin->attach()

aleksander0m commented:
> > The minimum flow for the DW5821e, which needs first fastboot, then qmi-pdc
> 
> So, what about this as a general rule: _if the plugin changes during detach_....
> 
> * run NewPlugin->write_firmware()
> * run NewPlugin->attach()
> * run OldPlugin->write_firmware()
> * run OldPlugin->attach()

For my use case it will be enough something like that I think, let me draft that.

hughsie commented:
> let me draft that

I'm happy to help if you want, fu_engine_install_blob() is an overgrown lump that's needed some love for some time.

aleksander0m commented:
> > let me draft that
> 
> I'm happy to help if you want, fu_engine_install_blob() is an overgrown lump that's needed some love for some time.

I've re-pushed the branch with the new logic now. For the additional write/attach procedures, I just copied the same code inside an additional if()... I was going to set that chunk of code as a separate method, but not sure how you would prefer to have that done.

This new branch works for the DW5821e and it should also work for other modems that just require fastboot without qmi-pdc and for those that require qmi-pdc without fastboot.

aleksander0m commented:
@hughsie I've rebased my branch on top of git master, and modified it to work with the new ANOTHER_WRITE_NEEDED flag, and it's working well. The MM device detach() method is a bit hacky, though, because I somehow need to know whether it's the first or second time called, and I rely on the type of FuMmDevice for that (i.e. whether it's created from MM or from udev events), but it works.

aleksander0m commented:
@hughsie should I edit this MR and add master as target branch? or, if you're happy with these changes, I could also force push it to your wip/hughsie/libmm branch?

hughsie commented:
I think rebasing against wip/hughsie/libmm and force pushing there is a good plan.

aleksander0m commented:
Ok, it's in wip/hughsie/libmm, so closing this MR.
